// @sap/cloud-sdk-core/dist/analytics/usage-analytics.js
"use strict";
/*!
 * Copyright (c) 2019 SAP SE or an SAP affiliate company. All rights reserved.
 */
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var axios_1 = __importDefault(require("axios"));
var perf_hooks_1 = require("perf_hooks");
var analytics_data_1 = require("./analytics-data");
var usageAnalyticsBlog = 'https://blogs.sap.com/2018/10/23/usage-analytics-s4sdk/';
var defaultURI = 'https://webanalytics.cfapps.eu10.hana.ondemand.com/tracker/log';
var defaultParameters = {
    action_name: 'SAP S/4HANA Cloud SDK',
    url: 'https://blogs.sap.com/2018/10/23/usage-analytics-s4sdk/',
    idsite: 'b67a7f90-dc52-72f0-cbac-18bf4147456a',
    idsitesub: 'test-jssdk',
    event_type: 'npm_start'
};
var defaultOptions = { useSalt: true, uri: defaultURI, idsitesub: defaultParameters.idsitesub, event_type: defaultParameters.event_type };
/**
 * Sends development environment data to SAP Web Analytic.
 * For detailed information, check the following blog post: https://blogs.sap.com/2018/10/23/usage-analytics-s4sdk/
 *
 * @param options Analytics options
 * @returns Promise
 * @hidden
 */
function sendAnalyticsData(options) {
    return __awaiter(this, void 0, void 0, function () {
        var data, params, before_1, sawParameters;
        return __generator(this, function (_a) {
            try {
                if (skipUsageAnalytics()) {
                    console.warn('Usage analytics is skipped.');
                    return [2 /*return*/];
                }
                if (!options) {
                    options = defaultOptions;
                }
                else {
                    Object.keys(defaultOptions).forEach(function (key) {
                        if (!options.hasOwnProperty(key)) {
                            options[key] = defaultOptions[key];
                        }
                    });
                }
                data = analytics_data_1.getAnalyticsData();
                params = {
                    project_id: analytics_data_1.getProjectIdentifier(options.useSalt),
                    os: Object.values(data.operatingSystemInfo).join(','),
                    node: data.nodeVersion,
                    npm: data.npmVersion,
                    typescript: data.withTypeScript.toString(),
                    sdk_dependencies: Object.keys(data.cloudSdkModules).join(','),
                    third_party_dependencies: Object.keys(data.thirdPartyDependencies).join(',')
                };
                logDescription();
                before_1 = perf_hooks_1.performance.now();
                sawParameters = getSAWParameters(options);
                return [2 /*return*/, axios_1.default
                        .get(options.uri, { params: __assign(__assign({}, sawParameters), payloadToCustomParameters(params)) })
                        .then(function () {
                        console.info('Usage data sent successfully.');
                        return perf_hooks_1.performance.now() - before_1;
                    })
                        .catch(function () {
                        console.info('Failed to send usage data.');
                        return undefined;
                    })];
            }
            catch (e) {
                console.warn('An error has occurred while executing usageAnalytics() : ' + e);
                console.warn('Usage analytics is skipped.');
            }
            return [2 /*return*/];
        });
    });
}
exports.sendAnalyticsData = sendAnalyticsData;
/**
 * Creates SAP Web Analytics regular parameters.
 *
 * @param options Request options
 * @returns SAP Web Analytics regular parameters
 */
function getSAWParameters(options) {
    return {
        action_name: defaultParameters.action_name,
        url: defaultParameters.url,
        idsite: defaultParameters.idsite,
        idsitesub: options.idsitesub ? options.idsitesub : defaultParameters.idsitesub,
        event_type: options.event_type ? options.event_type : defaultParameters.event_type
    };
}
/**
 * Creates SAP Web Analytics custom parameters. It maps user data to an object format compliant with SAP Web Analytics requirements.
 * Example:
 * {
 *   project_id: "myProjectId",
 *   os: "myOsInfo",
 *   node: "nodeVersion",
 *   npm: "npmVersion",
 *   typescript: "false",
 *   sdk_dependencies: "myCloudDependencies",
 *   third_party_dependencies: "myDependencies"
 * }
 * is converted to
 * {
 *   custom1: "project_id",
 *   e_a: "myProjectId",
 *   custom2: "os",
 *   e_2: "myOsInfo",
 *   custom3: "node",
 *   e_3: "nodeVersion",
 *   custom4: "npm",
 *   e_4: "npmVersion",
 *   custom5: "typescript",
 *   e_5: "false",
 *   custom6: "sdk_dependencies",
 *   e_6: "myCloudDependencies",
 *   custom7: "third_party_dependencies",
 *   e_7: "myDependencies"
 * }
 *
 * Notice: Properties order of the passed parameters is relevant in determining the values of the mapped object.
 *
 * @param params User data
 * @returns SAP Web Analytics custom parameters
 */
function payloadToCustomParameters(params) {
    var i = 1;
    return Object.entries(params).reduce(function (prev, _a) {
        var key = _a[0], value = _a[1];
        prev['custom' + i] = key;
        prev['e_' + (i === 1 ? 'a' : i)] = value;
        i++;
        return prev;
    }, {});
}
function logDescription() {
    console.info("SAP Cloud SDK - Usage Analytics\n  Thank you for contributing to our anonymized usage statistics. This allows us to improve the SAP Cloud SDK based on your usage.\n  We respect your privacy and intellectual property. Therefore, we only collect non-sensitive data about the use of the SDK. We do not collect personal information or data about the inner workings of your project. If you prefer to opt out or want to learn more, visit: " + usageAnalyticsBlog);
}
function skipUsageAnalytics() {
    return process.env.npm_config__sap_skip_usage_analytics && process.env.npm_config__sap_skip_usage_analytics.toLowerCase() === 'true';
}
