// @sap/cds-compiler/lib/gen/JSONLexer.js
// Generated from JSON.g4 by ANTLR 4.7.1
// jshint ignore: start
var antlr4 = require('antlr4/index');



  function getLoc(ctx,ret) {
    if(ctx.start) {
      ret.start = {
        offset:ctx.start.start,
        start: ctx.start.start,
        stop: ctx.start.stop,
        line: ctx.start.line,
        column: ctx.start.column + 1
      }
    }
    if(ctx.stop) {
      ret.stop = {
        offset: ctx.stop.stop+1,
        start: ctx.stop.start,
        stop: ctx.stop.stop,
        line: ctx.stop.line,
        column: ctx.stop.column + 1 + (ctx.stop.stop - ctx.stop.start) // end of last token
      }
    }
  } // getLoc

  function unquoteNumber(s) {
    return JSON.parse(s); // unquote
  }

  function unquoteString(s) {
    if(s.startsWith("'") && s.endsWith("'")) {
       let ss = s.substring(1,s.length-1);
       ss = ss.replace(/\\/g, '\\\\')
       ss = ss.replace(/\"/g, '\\"')
       s = '"' + ss + '"';
    }
    if(!s.startsWith('"') && !s.endsWith('"') && !isCharNumber(s[0])) { // plain string
      s = '"' + s + '"';
    }
    return JSON.parse(s); // unquote

    function isCharNumber(c) {
      return c >= '0' && c <= '9';
    }
  }



var serializedATN = ["\u0003\u608b\ua72a\u8133\ub9ed\u417c\u3be7\u7786\u5964",
    "\u0002\u0010\u00ed\b\u0001\u0004\u0002\t\u0002\u0004\u0003\t\u0003\u0004",
    "\u0004\t\u0004\u0004\u0005\t\u0005\u0004\u0006\t\u0006\u0004\u0007\t",
    "\u0007\u0004\b\t\b\u0004\t\t\t\u0004\n\t\n\u0004\u000b\t\u000b\u0004",
    "\f\t\f\u0004\r\t\r\u0004\u000e\t\u000e\u0004\u000f\t\u000f\u0004\u0010",
    "\t\u0010\u0004\u0011\t\u0011\u0004\u0012\t\u0012\u0004\u0013\t\u0013",
    "\u0004\u0014\t\u0014\u0004\u0015\t\u0015\u0004\u0016\t\u0016\u0004\u0017",
    "\t\u0017\u0004\u0018\t\u0018\u0004\u0019\t\u0019\u0004\u001a\t\u001a",
    "\u0004\u001b\t\u001b\u0004\u001c\t\u001c\u0004\u001d\t\u001d\u0004\u001e",
    "\t\u001e\u0004\u001f\t\u001f\u0004 \t \u0003\u0002\u0003\u0002\u0003",
    "\u0003\u0003\u0003\u0003\u0004\u0003\u0004\u0003\u0005\u0003\u0005\u0003",
    "\u0006\u0003\u0006\u0003\u0007\u0003\u0007\u0003\b\u0003\b\u0003\b\u0003",
    "\b\u0003\b\u0003\t\u0003\t\u0003\t\u0003\t\u0003\t\u0003\t\u0003\n\u0003",
    "\n\u0003\n\u0003\n\u0003\n\u0003\u000b\u0003\u000b\u0003\u000b\u0005",
    "\u000ba\n\u000b\u0003\f\u0003\f\u0003\f\u0003\f\u0007\fg\n\f\f\f\u000e",
    "\fj\u000b\f\u0003\r\u0003\r\u0003\r\u0003\r\u0003\r\u0003\r\u0005\r",
    "r\n\r\u0003\u000e\u0003\u000e\u0003\u000f\u0003\u000f\u0003\u0010\u0003",
    "\u0010\u0003\u0011\u0003\u0011\u0003\u0012\u0003\u0012\u0003\u0013\u0003",
    "\u0013\u0003\u0014\u0003\u0014\u0003\u0015\u0003\u0015\u0003\u0016\u0003",
    "\u0016\u0003\u0016\u0007\u0016\u0087\n\u0016\f\u0016\u000e\u0016\u008a",
    "\u000b\u0016\u0003\u0016\u0003\u0016\u0003\u0017\u0003\u0017\u0003\u0017",
    "\u0007\u0017\u0091\n\u0017\f\u0017\u000e\u0017\u0094\u000b\u0017\u0003",
    "\u0017\u0003\u0017\u0003\u0018\u0003\u0018\u0003\u0018\u0005\u0018\u009b",
    "\n\u0018\u0003\u0019\u0003\u0019\u0003\u0019\u0003\u0019\u0003\u0019",
    "\u0003\u0019\u0003\u001a\u0003\u001a\u0003\u001b\u0005\u001b\u00a6\n",
    "\u001b\u0003\u001b\u0003\u001b\u0003\u001b\u0006\u001b\u00ab\n\u001b",
    "\r\u001b\u000e\u001b\u00ac\u0003\u001b\u0005\u001b\u00b0\n\u001b\u0003",
    "\u001b\u0005\u001b\u00b3\n\u001b\u0003\u001b\u0003\u001b\u0003\u001b",
    "\u0003\u001b\u0005\u001b\u00b9\n\u001b\u0003\u001b\u0005\u001b\u00bc",
    "\n\u001b\u0003\u001c\u0003\u001c\u0003\u001c\u0007\u001c\u00c1\n\u001c",
    "\f\u001c\u000e\u001c\u00c4\u000b\u001c\u0005\u001c\u00c6\n\u001c\u0003",
    "\u001d\u0003\u001d\u0005\u001d\u00ca\n\u001d\u0003\u001d\u0003\u001d",
    "\u0003\u001e\u0006\u001e\u00cf\n\u001e\r\u001e\u000e\u001e\u00d0\u0003",
    "\u001e\u0003\u001e\u0003\u001f\u0003\u001f\u0003\u001f\u0003\u001f\u0007",
    "\u001f\u00d9\n\u001f\f\u001f\u000e\u001f\u00dc\u000b\u001f\u0003\u001f",
    "\u0003\u001f\u0003\u001f\u0003\u001f\u0003\u001f\u0003 \u0003 \u0003",
    " \u0003 \u0007 \u00e7\n \f \u000e \u00ea\u000b \u0003 \u0003 \u0003",
    "\u00da\u0002!\u0003\u0003\u0005\u0004\u0007\u0005\t\u0006\u000b\u0007",
    "\r\b\u000f\t\u0011\n\u0013\u000b\u0015\f\u0017\u0002\u0019\u0002\u001b",
    "\u0002\u001d\u0002\u001f\u0002!\u0002#\u0002%\u0002\'\u0002)\u0002+",
    "\u0002-\u0002/\u00021\u00023\u00025\r7\u00029\u0002;\u000e=\u000f?\u0010",
    "\u0003\u0002\u000e\u0007\u0002##%(--`a\u0080\u0080\u0004\u0002C\\c|",
    "\u0003\u00022;\u0004\u0002))^^\u0004\u0002$$^^\n\u0002$$11^^ddhhppt",
    "tvv\u0005\u00022;CHch\u0003\u00023;\u0004\u0002GGgg\u0004\u0002--//",
    "\u0005\u0002\u000b\f\u000f\u000f\"\"\u0004\u0002\f\f\u000f\u000f\u0002",
    "\u00f7\u0002\u0003\u0003\u0002\u0002\u0002\u0002\u0005\u0003\u0002\u0002",
    "\u0002\u0002\u0007\u0003\u0002\u0002\u0002\u0002\t\u0003\u0002\u0002",
    "\u0002\u0002\u000b\u0003\u0002\u0002\u0002\u0002\r\u0003\u0002\u0002",
    "\u0002\u0002\u000f\u0003\u0002\u0002\u0002\u0002\u0011\u0003\u0002\u0002",
    "\u0002\u0002\u0013\u0003\u0002\u0002\u0002\u0002\u0015\u0003\u0002\u0002",
    "\u0002\u00025\u0003\u0002\u0002\u0002\u0002;\u0003\u0002\u0002\u0002",
    "\u0002=\u0003\u0002\u0002\u0002\u0002?\u0003\u0002\u0002\u0002\u0003",
    "A\u0003\u0002\u0002\u0002\u0005C\u0003\u0002\u0002\u0002\u0007E\u0003",
    "\u0002\u0002\u0002\tG\u0003\u0002\u0002\u0002\u000bI\u0003\u0002\u0002",
    "\u0002\rK\u0003\u0002\u0002\u0002\u000fM\u0003\u0002\u0002\u0002\u0011",
    "R\u0003\u0002\u0002\u0002\u0013X\u0003\u0002\u0002\u0002\u0015`\u0003",
    "\u0002\u0002\u0002\u0017b\u0003\u0002\u0002\u0002\u0019q\u0003\u0002",
    "\u0002\u0002\u001bs\u0003\u0002\u0002\u0002\u001du\u0003\u0002\u0002",
    "\u0002\u001fw\u0003\u0002\u0002\u0002!y\u0003\u0002\u0002\u0002#{\u0003",
    "\u0002\u0002\u0002%}\u0003\u0002\u0002\u0002\'\u007f\u0003\u0002\u0002",
    "\u0002)\u0081\u0003\u0002\u0002\u0002+\u0083\u0003\u0002\u0002\u0002",
    "-\u008d\u0003\u0002\u0002\u0002/\u0097\u0003\u0002\u0002\u00021\u009c",
    "\u0003\u0002\u0002\u00023\u00a2\u0003\u0002\u0002\u00025\u00bb\u0003",
    "\u0002\u0002\u00027\u00c5\u0003\u0002\u0002\u00029\u00c7\u0003\u0002",
    "\u0002\u0002;\u00ce\u0003\u0002\u0002\u0002=\u00d4\u0003\u0002\u0002",
    "\u0002?\u00e2\u0003\u0002\u0002\u0002AB\u0007}\u0002\u0002B\u0004\u0003",
    "\u0002\u0002\u0002CD\u0007.\u0002\u0002D\u0006\u0003\u0002\u0002\u0002",
    "EF\u0007\u007f\u0002\u0002F\b\u0003\u0002\u0002\u0002GH\u0007<\u0002",
    "\u0002H\n\u0003\u0002\u0002\u0002IJ\u0007]\u0002\u0002J\f\u0003\u0002",
    "\u0002\u0002KL\u0007_\u0002\u0002L\u000e\u0003\u0002\u0002\u0002MN\u0007",
    "v\u0002\u0002NO\u0007t\u0002\u0002OP\u0007w\u0002\u0002PQ\u0007g\u0002",
    "\u0002Q\u0010\u0003\u0002\u0002\u0002RS\u0007h\u0002\u0002ST\u0007c",
    "\u0002\u0002TU\u0007n\u0002\u0002UV\u0007u\u0002\u0002VW\u0007g\u0002",
    "\u0002W\u0012\u0003\u0002\u0002\u0002XY\u0007p\u0002\u0002YZ\u0007w",
    "\u0002\u0002Z[\u0007n\u0002\u0002[\\\u0007n\u0002\u0002\\\u0014\u0003",
    "\u0002\u0002\u0002]a\u0005+\u0016\u0002^a\u0005-\u0017\u0002_a\u0005",
    "\u0017\f\u0002`]\u0003\u0002\u0002\u0002`^\u0003\u0002\u0002\u0002`",
    "_\u0003\u0002\u0002\u0002a\u0016\u0003\u0002\u0002\u0002bh\u0005\u0019",
    "\r\u0002cg\u0005\u0019\r\u0002dg\u0005)\u0015\u0002eg\u0005%\u0013\u0002",
    "fc\u0003\u0002\u0002\u0002fd\u0003\u0002\u0002\u0002fe\u0003\u0002\u0002",
    "\u0002gj\u0003\u0002\u0002\u0002hf\u0003\u0002\u0002\u0002hi\u0003\u0002",
    "\u0002\u0002i\u0018\u0003\u0002\u0002\u0002jh\u0003\u0002\u0002\u0002",
    "kr\u0005\u001d\u000f\u0002lr\u0005\u001b\u000e\u0002mr\u0005\u001f\u0010",
    "\u0002nr\u0005!\u0011\u0002or\u0005#\u0012\u0002pr\u0005\'\u0014\u0002",
    "qk\u0003\u0002\u0002\u0002ql\u0003\u0002\u0002\u0002qm\u0003\u0002\u0002",
    "\u0002qn\u0003\u0002\u0002\u0002qo\u0003\u0002\u0002\u0002qp\u0003\u0002",
    "\u0002\u0002r\u001a\u0003\u0002\u0002\u0002st\t\u0002\u0002\u0002t\u001c",
    "\u0003\u0002\u0002\u0002uv\t\u0003\u0002\u0002v\u001e\u0003\u0002\u0002",
    "\u0002wx\u0007,\u0002\u0002x \u0003\u0002\u0002\u0002yz\u0007A\u0002",
    "\u0002z\"\u0003\u0002\u0002\u0002{|\u00070\u0002\u0002|$\u0003\u0002",
    "\u0002\u0002}~\u0007/\u0002\u0002~&\u0003\u0002\u0002\u0002\u007f\u0080",
    "\u0007B\u0002\u0002\u0080(\u0003\u0002\u0002\u0002\u0081\u0082\t\u0004",
    "\u0002\u0002\u0082*\u0003\u0002\u0002\u0002\u0083\u0088\u0007)\u0002",
    "\u0002\u0084\u0087\u0005/\u0018\u0002\u0085\u0087\n\u0005\u0002\u0002",
    "\u0086\u0084\u0003\u0002\u0002\u0002\u0086\u0085\u0003\u0002\u0002\u0002",
    "\u0087\u008a\u0003\u0002\u0002\u0002\u0088\u0086\u0003\u0002\u0002\u0002",
    "\u0088\u0089\u0003\u0002\u0002\u0002\u0089\u008b\u0003\u0002\u0002\u0002",
    "\u008a\u0088\u0003\u0002\u0002\u0002\u008b\u008c\u0007)\u0002\u0002",
    "\u008c,\u0003\u0002\u0002\u0002\u008d\u0092\u0007$\u0002\u0002\u008e",
    "\u0091\u0005/\u0018\u0002\u008f\u0091\n\u0006\u0002\u0002\u0090\u008e",
    "\u0003\u0002\u0002\u0002\u0090\u008f\u0003\u0002\u0002\u0002\u0091\u0094",
    "\u0003\u0002\u0002\u0002\u0092\u0090\u0003\u0002\u0002\u0002\u0092\u0093",
    "\u0003\u0002\u0002\u0002\u0093\u0095\u0003\u0002\u0002\u0002\u0094\u0092",
    "\u0003\u0002\u0002\u0002\u0095\u0096\u0007$\u0002\u0002\u0096.\u0003",
    "\u0002\u0002\u0002\u0097\u009a\u0007^\u0002\u0002\u0098\u009b\t\u0007",
    "\u0002\u0002\u0099\u009b\u00051\u0019\u0002\u009a\u0098\u0003\u0002",
    "\u0002\u0002\u009a\u0099\u0003\u0002\u0002\u0002\u009b0\u0003\u0002",
    "\u0002\u0002\u009c\u009d\u0007w\u0002\u0002\u009d\u009e\u00053\u001a",
    "\u0002\u009e\u009f\u00053\u001a\u0002\u009f\u00a0\u00053\u001a\u0002",
    "\u00a0\u00a1\u00053\u001a\u0002\u00a12\u0003\u0002\u0002\u0002\u00a2",
    "\u00a3\t\b\u0002\u0002\u00a34\u0003\u0002\u0002\u0002\u00a4\u00a6\u0007",
    "/\u0002\u0002\u00a5\u00a4\u0003\u0002\u0002\u0002\u00a5\u00a6\u0003",
    "\u0002\u0002\u0002\u00a6\u00a7\u0003\u0002\u0002\u0002\u00a7\u00a8\u0005",
    "7\u001c\u0002\u00a8\u00aa\u00070\u0002\u0002\u00a9\u00ab\t\u0004\u0002",
    "\u0002\u00aa\u00a9\u0003\u0002\u0002\u0002\u00ab\u00ac\u0003\u0002\u0002",
    "\u0002\u00ac\u00aa\u0003\u0002\u0002\u0002\u00ac\u00ad\u0003\u0002\u0002",
    "\u0002\u00ad\u00af\u0003\u0002\u0002\u0002\u00ae\u00b0\u00059\u001d",
    "\u0002\u00af\u00ae\u0003\u0002\u0002\u0002\u00af\u00b0\u0003\u0002\u0002",
    "\u0002\u00b0\u00bc\u0003\u0002\u0002\u0002\u00b1\u00b3\u0007/\u0002",
    "\u0002\u00b2\u00b1\u0003\u0002\u0002\u0002\u00b2\u00b3\u0003\u0002\u0002",
    "\u0002\u00b3\u00b4\u0003\u0002\u0002\u0002\u00b4\u00b5\u00057\u001c",
    "\u0002\u00b5\u00b6\u00059\u001d\u0002\u00b6\u00bc\u0003\u0002\u0002",
    "\u0002\u00b7\u00b9\u0007/\u0002\u0002\u00b8\u00b7\u0003\u0002\u0002",
    "\u0002\u00b8\u00b9\u0003\u0002\u0002\u0002\u00b9\u00ba\u0003\u0002\u0002",
    "\u0002\u00ba\u00bc\u00057\u001c\u0002\u00bb\u00a5\u0003\u0002\u0002",
    "\u0002\u00bb\u00b2\u0003\u0002\u0002\u0002\u00bb\u00b8\u0003\u0002\u0002",
    "\u0002\u00bc6\u0003\u0002\u0002\u0002\u00bd\u00c6\u00072\u0002\u0002",
    "\u00be\u00c2\t\t\u0002\u0002\u00bf\u00c1\t\u0004\u0002\u0002\u00c0\u00bf",
    "\u0003\u0002\u0002\u0002\u00c1\u00c4\u0003\u0002\u0002\u0002\u00c2\u00c0",
    "\u0003\u0002\u0002\u0002\u00c2\u00c3\u0003\u0002\u0002\u0002\u00c3\u00c6",
    "\u0003\u0002\u0002\u0002\u00c4\u00c2\u0003\u0002\u0002\u0002\u00c5\u00bd",
    "\u0003\u0002\u0002\u0002\u00c5\u00be\u0003\u0002\u0002\u0002\u00c68",
    "\u0003\u0002\u0002\u0002\u00c7\u00c9\t\n\u0002\u0002\u00c8\u00ca\t\u000b",
    "\u0002\u0002\u00c9\u00c8\u0003\u0002\u0002\u0002\u00c9\u00ca\u0003\u0002",
    "\u0002\u0002\u00ca\u00cb\u0003\u0002\u0002\u0002\u00cb\u00cc\u00057",
    "\u001c\u0002\u00cc:\u0003\u0002\u0002\u0002\u00cd\u00cf\t\f\u0002\u0002",
    "\u00ce\u00cd\u0003\u0002\u0002\u0002\u00cf\u00d0\u0003\u0002\u0002\u0002",
    "\u00d0\u00ce\u0003\u0002\u0002\u0002\u00d0\u00d1\u0003\u0002\u0002\u0002",
    "\u00d1\u00d2\u0003\u0002\u0002\u0002\u00d2\u00d3\b\u001e\u0002\u0002",
    "\u00d3<\u0003\u0002\u0002\u0002\u00d4\u00d5\u00071\u0002\u0002\u00d5",
    "\u00d6\u0007,\u0002\u0002\u00d6\u00da\u0003\u0002\u0002\u0002\u00d7",
    "\u00d9\u000b\u0002\u0002\u0002\u00d8\u00d7\u0003\u0002\u0002\u0002\u00d9",
    "\u00dc\u0003\u0002\u0002\u0002\u00da\u00db\u0003\u0002\u0002\u0002\u00da",
    "\u00d8\u0003\u0002\u0002\u0002\u00db\u00dd\u0003\u0002\u0002\u0002\u00dc",
    "\u00da\u0003\u0002\u0002\u0002\u00dd\u00de\u0007,\u0002\u0002\u00de",
    "\u00df\u00071\u0002\u0002\u00df\u00e0\u0003\u0002\u0002\u0002\u00e0",
    "\u00e1\b\u001f\u0002\u0002\u00e1>\u0003\u0002\u0002\u0002\u00e2\u00e3",
    "\u00071\u0002\u0002\u00e3\u00e4\u00071\u0002\u0002\u00e4\u00e8\u0003",
    "\u0002\u0002\u0002\u00e5\u00e7\n\r\u0002\u0002\u00e6\u00e5\u0003\u0002",
    "\u0002\u0002\u00e7\u00ea\u0003\u0002\u0002\u0002\u00e8\u00e6\u0003\u0002",
    "\u0002\u0002\u00e8\u00e9\u0003\u0002\u0002\u0002\u00e9\u00eb\u0003\u0002",
    "\u0002\u0002\u00ea\u00e8\u0003\u0002\u0002\u0002\u00eb\u00ec\b \u0002",
    "\u0002\u00ec@\u0003\u0002\u0002\u0002\u0018\u0002`fhq\u0086\u0088\u0090",
    "\u0092\u009a\u00a5\u00ac\u00af\u00b2\u00b8\u00bb\u00c2\u00c5\u00c9\u00d0",
    "\u00da\u00e8\u0003\b\u0002\u0002"].join("");


var atn = new antlr4.atn.ATNDeserializer().deserialize(serializedATN);

var decisionsToDFA = atn.decisionToState.map( function(ds, index) { return new antlr4.dfa.DFA(ds, index); });

function JSONLexer(input) {
	antlr4.Lexer.call(this, input);
    this._interp = new antlr4.atn.LexerATNSimulator(this, atn, decisionsToDFA, new antlr4.PredictionContextCache());
    return this;
}

JSONLexer.prototype = Object.create(antlr4.Lexer.prototype);
JSONLexer.prototype.constructor = JSONLexer;

Object.defineProperty(JSONLexer.prototype, "atn", {
        get : function() {
                return atn;
        }
});

JSONLexer.EOF = antlr4.Token.EOF;
JSONLexer.T__0 = 1;
JSONLexer.T__1 = 2;
JSONLexer.T__2 = 3;
JSONLexer.T__3 = 4;
JSONLexer.T__4 = 5;
JSONLexer.T__5 = 6;
JSONLexer.T__6 = 7;
JSONLexer.T__7 = 8;
JSONLexer.T__8 = 9;
JSONLexer.STRING = 10;
JSONLexer.NUMBER = 11;
JSONLexer.WHITESPACE = 12;
JSONLexer.MULTI_LINE_COMMENT = 13;
JSONLexer.SINGLE_LINE_COMMENT = 14;

JSONLexer.prototype.channelNames = [ "DEFAULT_TOKEN_CHANNEL", "HIDDEN" ];

JSONLexer.prototype.modeNames = [ "DEFAULT_MODE" ];

JSONLexer.prototype.literalNames = [ null, "'{'", "','", "'}'", "':'", "'['", 
                                     "']'", "'true'", "'false'", "'null'" ];

JSONLexer.prototype.symbolicNames = [ null, null, null, null, null, null, 
                                      null, null, null, null, "STRING", 
                                      "NUMBER", "WHITESPACE", "MULTI_LINE_COMMENT", 
                                      "SINGLE_LINE_COMMENT" ];

JSONLexer.prototype.ruleNames = [ "T__0", "T__1", "T__2", "T__3", "T__4", 
                                  "T__5", "T__6", "T__7", "T__8", "STRING", 
                                  "ASTRING", "STRSYM", "SPECHAR", "LETTER", 
                                  "STAR", "QUESTION", "DOT", "MINUS", "AtSIGN", 
                                  "DIGIT", "Q1STRING", "Q2STRING", "ESC", 
                                  "UNICODE", "HEX", "NUMBER", "INT", "EXP", 
                                  "WHITESPACE", "MULTI_LINE_COMMENT", "SINGLE_LINE_COMMENT" ];

JSONLexer.prototype.grammarFileName = "JSON.g4";



exports.JSONLexer = JSONLexer;
