// @sap/cds-services/lib/services/utils/constants.js
const RO = ['READ']
const CU = ['CREATE', 'UPDATE']
const CD = ['CREATE', 'DELETE']
const UD = ['UPDATE', 'DELETE']
const CUD = ['CREATE', ...UD]
const CRUD = [...RO, ...CUD]
const DRAFT = ['CANCEL', 'PATCH', 'NEW', 'EDIT', 'SAVE', 'draftPrepare', 'draftActivate']
const RJ_DRAFT = ['draftActivate', 'EDIT', 'draftPrepare']
const BEFORE_DRAFT = [
  ['PATCH', 'beforeUpdateDraft'],
  ['UPDATE', 'beforeUpdateDraft'],
  ['CANCEL', 'beforeDeleteOrCancelDraft']
]
const ON_DRAFT = [
  ['READ', 'onReadDraft'],
  ['CREATE', 'onDraftActivate'],
  ['PATCH', 'onPatchDraft'],
  ['UPDATE', 'onDraftActivate'],
  ['CANCEL', 'onCancelDraft']
]
const MOD = ['CREATE', 'UPDATE', 'NEW', 'PATCH', 'PUT']

const PR_AC_DRAFT = [['draftPrepare', 'onDraftPrepare'], ['draftActivate', 'onDraftActivateEvent']]
const TRANSACTION = ['COMMIT', 'ROLLBACK']

const DRAFT_COLUMNS = ['IsActiveEntity', 'HasActiveEntity', 'HasDraftEntity', 'DraftAdministrativeData_DraftUUID']

module.exports = {
  messages: {
    DB_CONNECTION_MISSING: 'Database connection is missing'
  },
  events: {
    DEFAULT: [...CRUD, ...DRAFT, ...TRANSACTION],
    RO,
    CU,
    CD,
    UD,
    CUD,
    CRUD,
    DRAFT,
    CRUD_DRAFT: [...CRUD, ...DRAFT],
    CUD_DRAFT: [...CUD, ...DRAFT],
    RJ_DRAFT,
    BEFORE_DRAFT,
    ON_DRAFT,
    PR_AC_DRAFT,
    TRANSACTION,
    MOD
  },
  DRAFT_COLUMNS
}
