// @sap/cds-compiler/benchmark/jsonReporter.js
'use strict';

/* eslint no-console:off */

function JsonReporter(benchmarkSuite) {
  const printBenchmarks = () => {
    const output = {
      results: []
    };

    // only output system/git information if it's available
    if (benchmarkSuite.system) {
      output.system = benchmarkSuite.system;
    }
    if (benchmarkSuite.git) {
      output.git = benchmarkSuite.git;
    }

    for (let benchmark of benchmarkSuite.benchmarks) {
      output.results.push({
        name:             benchmark.identifier,
        description:      benchmark.description,
        warmUpExecutions: benchmark.options.warmUpExecutions,
        executions:       benchmark.options.executions,
        runtime:    timeInMilliSeconds(benchmark.runtime),
        avgRuntime: timeInMilliSeconds([
          benchmark.runtime[0] / benchmark.options.executions,
          benchmark.runtime[1] / benchmark.options.executions
        ])
      });
    }

    // Write pretty-printed JSON (two spaces indentation, no replacer)
    console.log(JSON.stringify(output, undefined, 2));
  };

  // Converts a process.hrtime value to a number in milliseconds.
  let timeInMilliSeconds = (hrtime) => {
    const [s, ns] = hrtime;
    const ms = (s * 1000 + ns / 1000 / 1000);
    return ms;
  };

  printBenchmarks();
}

module.exports = JsonReporter;
