// @sap/cds-services/lib/services/handlers/onCreateDraft.js
const onDraftActivate = require('./onDraftActivate')()
const { isNavigationToMany } = require('../utils/compositionTree')
const { removeDraftUUID, ensureDraftsSuffix } = require('../utils/draftUtils')
const {
  messages: { DB_CONNECTION_MISSING }
} = require('../utils/constants')

const _getUpdateDraftAdminCQN = ({ statements, user }, draftUUID) => {
  return statements
    .UPDATE('DRAFT.DraftAdministrativeData')
    .set({
      InProcessByUser: user.id,
      LastChangedByUser: user.id,
      LastChangeDateTime: new Date(Date.now()).toISOString()
    })
    .where({ DraftUUID: draftUUID })
}

const _getInsertDraftAdminCQN = ({ statements, user }, uuid) => {
  const time = new Date(Date.now()).toISOString()

  return statements.INSERT.into('DRAFT.DraftAdministrativeData').entries({
    DraftUUID: uuid,
    CreationDateTime: time,
    CreatedByUser: user.id,
    LastChangeDateTime: time,
    LastChangedByUser: user.id,
    DraftIsCreatedByMe: true,
    DraftIsProcessedByMe: true,
    InProcessByUser: user.id
  })
}

const _getInsertDataCQN = (context, draftUUID) => {
  const draftName = ensureDraftsSuffix(context.target.name)

  const insertData = context.statements.INSERT.into(draftName).entries(context.query.INSERT.entries[0]) // entries is always set because there are no entities without keys

  context.data.IsActiveEntity = false
  context.data.HasDraftEntity = false
  context.data.HasActiveEntity = false
  context.data.DraftAdministrativeData_DraftUUID = draftUUID

  return insertData
}
/**
 * Generic Handler for CREATE requests in the context of draft.
 * In case of success it returns the created entry.
 *
 * @alias module:handlers.onCreateDraft
 */
const onCreateDraft = () => context => {
  if (!context.isDraftChange) {
    return onDraftActivate(context)
  }

  if (!context.run) {
    context.log.warn(DB_CONNECTION_MISSING)
    return Promise.resolve(context.data)
  }

  const navigationToMany = isNavigationToMany(context)

  const adminDataCQN = navigationToMany
    ? _getUpdateDraftAdminCQN(context, context.data.DraftAdministrativeData_DraftUUID)
    : _getInsertDraftAdminCQN(context, context.data.DraftAdministrativeData_DraftUUID)
  const insertDataCQN = _getInsertDataCQN(context, context.data.DraftAdministrativeData_DraftUUID)

  return Promise.all([context.run(adminDataCQN), context.run(insertDataCQN)]).then(() => {
    return removeDraftUUID(Object.assign({}, context.data))
  })
}

module.exports = onCreateDraft
